### DISCUSSION TITLE
datacoord cannot start anymore, etcd cluster issue ?

### DISCUSSION THREAD

**Discussion by Archalbc** [milvus-io/milvus#27597] [Q&A and General discussion]
Hi!

My datacoord cannot start anymore, sound like there is an issue with the etcd cluster but I cannot understand why.
My etcd cluster seems healthy using EBS gp3 volume type with enough CPU (limit 8 )
    ebs.csi.aws.com/iops: "6000"
    ebs.csi.aws.com/throughput: "125"
    ebs.csi.aws.com/volumeType: gp3

![image (10)](https://github.com/milvus-io/milvus/assets/144429952/7ac21fd6-c021-4d7f-b5e5-3110457de94e)

* 1.80M keys
* DBsize 635 MB

Cluster still not used in production.

but on the datacoord logs there are lines like that :
```
[2023/10/09 15:00:45.531 +00:00] [DEBUG] [sessionutil/session_util.go:242] ["Session connect to etcd success"]                                                                                                     
[2023/10/09 15:00:45.603 +00:00] [INFO] [storage/minio_chunk_manager.go:154] ["minio chunk manager init success."] [bucketname=TRUNCATED] [root=milvus]       
[2023/10/09 15:00:50.414 +00:00] [WARN] [etcd/etcd_kv.go:647] ["Slow etcd operation(WalkWithPagination)"] ["time spent"=4.810268369s] [prefix=milvus-etcd/meta/datacoord-meta/s]                                   
[2023/10/09 15:00:50.424 +00:00] [DEBUG] [config/etcd_source.go:141] ["etcd refreshConfigurations"] [prefix=milvus-etcd/config] [endpoints="[milvus-etcd:2379]"]
[2023/10/09 15:00:55.424 +00:00] [DEBUG] [config/etcd_source.go:141] ["etcd refreshConfigurations"] [prefix=milvus-etcd/config] [endpoints="[milvus-etcd:2379]"]
{"level":"warn","ts":"2023-10-09T15:00:55.605Z","logger":"etcd-client","caller":"v3@v3.5.5/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc000da8700/milvus-etcd:23
79","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
[2023/10/09 15:00:55.605 +00:00] [ERROR] [datacoord/server.go:547] ["retry func failed"] ["retry time"=0] [error="context deadline exceeded"] [stack="github.com/milvus-io/milvus/internal/datacoord.(*Server).init
Meta\n\t/go/src/github.com/milvus-io/milvus/internal/datacoord/server.go:547\ngithub.com/milvus-io/milvus/internal/datacoord.(*Server).initDataCoord\n\t/go/src/github.com/milvus-io/milvus/internal/datacoord/serv
er.go:333\ngithub.com/milvus-io/milvus/internal/datacoord.(*Server).Init\n\t/go/src/github.com/milvus-io/milvus/internal/datacoord/server.go:316\ngithub.com/milvus-io/milvus/internal/distributed/datacoord.(*Serv
er).init\n\t/go/src/github.com/milvus-io/milvus/internal/distributed/datacoord/service.go:108\ngithub.com/milvus-io/milvus/internal/distributed/datacoord.(*Server).Run\n\t/go/src/github.com/milvus-io/milvus/inte
rnal/distributed/datacoord/service.go:229\ngithub.com/milvus-io/milvus/cmd/components.(*DataCoord).Run\n\t/go/src/github.com/milvus-io/milvus/cmd/components/data_coord.go:49\ngithub.com/milvus-io/milvus/cmd/role
s.runComponent[...].func1\n\t/go/src/github.com/milvus-io/milvus/cmd/roles/roles.go:112"]
```

and on the etcd logs:
```
{"level":"info","ts":"2023-10-09T15:06:36.299Z","caller":"traceutil/trace.go:171","msg":"trace[687297061] range","detail":"{range_begin:milvus-etcd/meta/datacoord-meta/binlog/444749944847657311/444749944847657312/444749944862769784/111\u0000; range_end:milvus-etcd/meta/datacoord-meta/binloh; response_count:2000; response_revision:5604002; }","duration":"153.292827ms","start":"2023-10-09T15:06:36.145Z","end":"2023-10-09T15:06:36.299Z","steps":["trace[687297061] 'range keys from in-memory index tree'  (duration: 150.614707ms)"],"step_count":1}
{"level":"warn","ts":"2023-10-09T15:06:36.456Z","caller":"etcdserver/util.go:166","msg":"apply request took too long","took":"151.985268ms","expected-duration":"100ms","prefix":"read-only range ","request":"key:\"milvus-etcd/meta/datacoord-meta/binlog/444749944847657311/444749944847657312/444749944862780777/109\\000\" range_end:\"milvus-etcd/meta/datacoord-meta/binloh\" limit:2000 ","response":"range_response_count:2000 size:311521"}
{"level":"info","ts":"2023-10-09T15:06:36.456Z","caller":"traceutil/trace.go:171","msg":"trace[25318916] range","detail":"{range_begin:milvus-etcd/meta/datacoord-meta/binlog/444749944847657311/444749944847657312/444749944862780777/109\u0000; range_end:milvus-etcd/meta/datacoord-meta/binloh; response_count:2000; response_revision:5604002; }","duration":"152.045724ms","start":"2023-10-09T15:06:36.304Z","end":"2023-10-09T15:06:36.456Z","steps":["trace[25318916] 'range keys from in-memory index tree'  (duration: 149.482152ms)"],"step_count":1}
{"level":"warn","ts":"2023-10-09T15:06:36.610Z","caller":"etcdserver/util.go:166","msg":"apply request took too long","took":"149.009803ms","expected-duration":"100ms","prefix":"read-only range ","request":"key:\"milvus-etcd/meta/datacoord-meta/binlog/444749944847657311/444749944847657312/444749944862785104/107\\000\" range_end:\"milvus-etcd/meta/datacoord-meta/binloh\" limit:2000 ","response":"range_response_count:2000 size:311461"}
```

Thank you for your help (logs in attached piece).

[etcd.log](https://github.com/milvus-io/milvus/files/12854138/etcd.log)
[datacoord.log](https://github.com/milvus-io/milvus/files/12854139/datacoord.log)

**Comment by yhmo** [comment]
Do you have etcd log between 15:00:00 and 15:02:00 ? There are several warnings "Slow etcd operation(WalkWithPagination)" during this time. "Slow etcd operation" means etcd api latency is very high(2~5 seconds).

**Reply by Archalbc** [reply]
Hello,

I did a "rollout restart sts" to get fresh logs from etcd nodes.

You can find them in attached pieces.

Thank you
[milvus-etcd-0.log](https://github.com/milvus-io/milvus/files/12855758/milvus-etcd-0.log)
[milvus-etcd-1.log](https://github.com/milvus-io/milvus/files/12855759/milvus-etcd-1.log)
[milvus-etcd-2.log](https://github.com/milvus-io/milvus/files/12855756/milvus-etcd-2.log)

**Reply by yhmo** [reply]
In the milvus-etcd-1.log, you can see there are thousands of warn messages like this:
`{"level":"warn","ts":"2023-10-10T09:52:57.020Z","caller":"etcdserver/util.go:166","msg":"apply request took too long","took":"148.605485ms","expected-duration":"100ms","prefix":"read-only range ","request":"key:\"milvus-etcd/meta/datacoord-meta/binlog/444749944847657311/444749944847657312/444749944862793546/103\\000\" range_end:\"milvus-etcd/meta/datacoord-meta/binloh\" limit:2000 ","response":"range_response_count:2000 size:311554"}`

If you search "etcd apply request took too long" on google, you will see the answer:
`If the average apply duration exceeds 100 milliseconds, etcd will warn that entries are taking too long to apply. Usually this issue is caused by a slow disk. The disk could be experiencing contention among etcd and other applications, or the disk is too simply slow (e.g., a shared virtualized disk).`

In our document, the prerequisitepage: https://milvus.io/docs/prerequisite-docker.md
etcd	3.5.0	See [additional disk requirements](https://milvus.io/docs/prerequisite-docker.md#Additional-disk-requirements).
It mentioned "Disk performance is critical to etcd. It is highly recommended that you use local NVMe SSDs. Slower disk response may cause frequent cluster elections that will eventually degrade the etcd service."

Etcd is the key dependency service of milvus, if the etcd is not running on high performance disk, milvus could run into unexpected problems such timeout or deadline exceed.

**Comment by Archalbc** [comment]
Hello,

I was able to test my disk, results seems pretty good :o
```
/data # fio --rw=write --ioengine=sync --fdatasync=1 --directory=test-data --size=2200m --bs=2300 --name=mytest
mytest: (g=0): rw=write, bs=(R) 2300B-2300B, (W) 2300B-2300B, (T) 2300B-2300B, ioengine=sync, iodepth=1
fio-3.35-137-g16b9f2
Starting 1 process
mytest: Laying out IO file (1 file / 2200MiB)
Jobs: 1 (f=1): [W(1)][63.9%][w=1459KiB/s][w=650 IOPS][eta 08m:58s]
Jobs: 1 (f=1): [W(1)][100.0%][w=1547KiB/s][w=689 IOPS][eta 00m:00s]
mytest: (groupid=0, jobs=1): err= 0: pid=53: Thu Oct 12 14:30:47 2023
  write: IOPS=681, BW=1532KiB/s (1568kB/s)(2200MiB/1470798msec); 0 zone resets
    clat (nsec): min=1470, max=1094.9k, avg=4868.29, stdev=2763.00
     lat (nsec): min=1521, max=1095.4k, avg=4962.57, stdev=2796.34
    clat percentiles (nsec):
     |  1.00th=[ 1960],  5.00th=[ 2288], 10.00th=[ 2576], 20.00th=[ 3344],
     | 30.00th=[ 3760], 40.00th=[ 4080], 50.00th=[ 4384], 60.00th=[ 4768],
     | 70.00th=[ 5216], 80.00th=[ 5984], 90.00th=[ 7328], 95.00th=[ 8768],
     | 99.00th=[13888], 99.50th=[16192], 99.90th=[22656], 99.95th=[25984],
     | 99.99th=[36096]
   bw (  KiB/s): min= 1024, max= 1711, per=99.96%, avg=1531.65, stdev=102.37, samples=2941
   iops        : min=  456, max=  762, avg=682.11, stdev=45.56, samples=2941
  lat (usec)   : 2=1.28%, 4=36.37%, 10=59.02%, 20=3.14%, 50=0.18%
  lat (usec)   : 100=0.01%, 250=0.01%, 500=0.01%, 750=0.01%
  lat (msec)   : 2=0.01%
  fsync/fdatasync/sync_file_range:
    sync (usec): min=412, max=45411, avg=1460.62, stdev=755.34
    sync percentiles (usec):
     |  1.00th=[  478],  5.00th=[  523], 10.00th=[  562], 20.00th=[  635],
     | 30.00th=[  725], 40.00th=[  816], 50.00th=[ 1795], 60.00th=[ 1926],
     | 70.00th=[ 2024], 80.00th=[ 2147], 90.00th=[ 2278], 95.00th=[ 2409],
     | 99.00th=[ 2802], 99.50th=[ 3064], 99.90th=[ 4424], 99.95th=[ 5145],
     | 99.99th=[ 8717]
  cpu          : usr=0.20%, sys=1.84%, ctx=2118558, majf=0, minf=11
  IO depths    : 1=200.0%, 2=0.0%, 4=0.0%, 8=0.0%, 16=0.0%, 32=0.0%, >=64=0.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     issued rwts: total=0,1002985,0,0 short=1002985,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=1

Run status group 0 (all jobs):
  WRITE: bw=1532KiB/s (1568kB/s), 1532KiB/s-1532KiB/s (1568kB/s-1568kB/s), io=2200MiB (2307MB), run=1470798-1470798msec

Disk stats (read/write):
  nvme1n1: ios=0/2129522, sectors=0/30543856, merge=0/1688462, ticks=0/1432875, in_queue=1432875, util=100.00%```

```

fsync 99th at 2.8 ms.

I can see from this graph : 
![image](https://github.com/milvus-io/milvus/assets/144429952/915d4d2e-e98d-4fe9-ae7b-52fc34e4a8b2)

The AVG is 4ms, which is good I think.
But something happened a 11h, compaction rate spike ?

pod resources:
```
    Limits:                                                                                                                                                                                                        
      cpu:     6                                                                                                                                                                                                   
      memory:  8Gi                                                                                                                                                                                                 
    Requests:                                                                                                                                                                                                      
      cpu:      2                                                                                                                                                                                                  
      memory:   4Gi     
```

I had reset my cluster several time to replicate this behavior, "slow" logs appears near 1.8M of keys inside etcd cluster.

The graph "Slow operation" indeed displays our issue (attached piece).
![screenshot-1](https://github.com/milvus-io/milvus/assets/144429952/69984701-65b4-4e51-bf45-21d23d911454)
![screenshot-2](https://github.com/milvus-io/milvus/assets/144429952/95d94c9f-0559-434c-bff7-b76a12323e48)
![screenshot-3](https://github.com/milvus-io/milvus/assets/144429952/52c5c52b-4bed-4c8f-9412-aeb06177669a)
![screenshot-4](https://github.com/milvus-io/milvus/assets/144429952/31b6fd3d-8433-4221-90d3-f2b3cfaa3c68)

**Reply by yhmo** [reply]
The compaction is etcd internal compaction, not milvus compaction.

In the etcd guideline, this page is for disk recommendation: https://etcd.io/docs/v3.5/op-guide/hardware/#disks
It says:
- For heavily loaded clusters, 500 sequential IOPS is recommended
- concurrent IOPS can be 10x greater than the sequential IOPS
- To measure actual sequential IOPS, we suggest using a disk benchmarking tool such as [diskbench](https://github.com/ongardie/diskbenchmark) or [fio](https://github.com/axboe/fio)


Some suggestions:
- make sure the etcd disk is SSD
- upgrade to 2.2.14(because we did some optimization for etcd usage)
- don't call flush() every time after insert()
- try inserting vectors batch by batch, instead of one by one

**Comment by Archalbc** [comment]
Hi,

I did some tests on PVC and even benchmarks using VectorDBBench.
It really seems nothing is wrong with my ETCD cluster
fsync average are pretty low while inserting 17K/s vectors.
![screenshot-1](https://github.com/milvus-io/milvus/assets/144429952/9489c71b-8cea-43a8-8104-3a54859887c9)
![screenshot-2](https://github.com/milvus-io/milvus/assets/144429952/1757ba17-d6a1-4464-81a4-2e34647e1fd7)

But if I restart the datacoord, slow operation appears while it's trying to access to a specific key/operations, here is the error:
```
{"level":"warn","ts":"2023-10-19T15:28:43.063Z","caller":"v3rpc/interceptor.go:197","msg":"request stats","start time":"2023-10-19T15:28:42.729Z","time spent":"333.52056ms","remote":"10.157.65.123:54432","response type":"/etcdserverpb.KV/Range","request count":0,"request size":145,"response count":1608575,"response size":307474,"request content":"key:\"milvus-etcd/meta/datacoord-meta/binlog/445003311460385739/445003311460385740/445019475471386423/102\\000\" range_end:\"milvus-etcd/meta/datacoord-meta/binloh\" limit:2000 "}
{"level":"warn","ts":"2023-10-19T15:28:43.063Z","caller":"etcdserver/util.go:166","msg":"apply request took too long","took":"333.422662ms","expected-duration":"100ms","prefix":"read-only range ","request":"key:\"milvus-etcd/meta/datacoord-meta/binlog/445003311460385739/445003311460385740/445019475471386423/102\\000\" range_end:\"milvus-etcd/meta/datacoord-meta/binloh\" limit:2000 ","response":"range_response_count:2000 size:307452"}
```

it always the same kind of log for the same kind of requested key.
Don't you think there is an issue with this datacoord request ? Is it only performed at the datacoord's boot ?

Thank you for you help.

**Reply by yhmo** [reply]
Still, the etcd warning: "apply request took too long","took":"333.422662ms"
The "response size":307474 is normal, not a huge response, but the latency is 300ms+. 
Could be network issue? I am not sure.

**Reply by Archalbc** [reply]
Hi,

Maybe you are right, here is the network usage for datacoord.
![datacoord](https://github.com/milvus-io/milvus/assets/144429952/875c5135-f319-4fff-b3c3-d00873534aea)
(datacoord restarted around 17h00)

etcd network
![etcd network](https://github.com/milvus-io/milvus/assets/144429952/8337b250-c887-4a31-a1ed-4f7c4bd05e6e)


I'm asking to the infrastructure team if they can check my pod's host metrics to see if I'm reaching a bandwith limitation.

Just for my understanding, could you tell me what the data coord does during its boot ? 
Apparently ETCD send a lot of stuff to a starting datacoord

The etcd cluster writes logs "apply request took too long" while he's sending stuff to the datacoord right ?

Thank you

**Reply by yhmo** [reply]
When Datacoord is initializing, it reads a lot of information from the ETCD, including:
- data channels info(each collection/partition has several channels according to the shards number)
- paths of data files for all the collections/segments(each collection has multiple segments, each segment has multiple data files)

So, if there are more collections/partitions, more segments, the datacoord will read more data from ETCD, more times to call ETCD interfaces. We are working on optimizing it by combining multiple requests into one call.

**Reply by Archalbc** [reply]
That's very good to know ! Thank !
I'll keep you up to date as soon as I have more info about some network issue.

**Reply by Archalbc** [reply]
Hi,

Here is some news : 
* Network is not reaching any bandwith limitations
* Disk I/O look fine (using EBS-GP3)
![etcd-disk-operations](https://github.com/milvus-io/milvus/assets/144429952/92616945-e4fe-41fe-9668-fa21239f9868)
![etcd-disk-sync-duration](https://github.com/milvus-io/milvus/assets/144429952/e335cfd2-8bf4-4f66-8723-def16a51a760)

It seems etcd disks are not under pressure but I can't really understand why it seems the network do not go beyond 10MB/s.
Eitherway, we still have no clue of what's is going on with some datacoord's requests...

Should we play with datacoord configuration like client/server GRPC size ?
https://milvus.io/docs/v2.0.x/configure_datacoord.md#dataCoordgrpcclientMaxRecvSize

I can see several logs on the datacoord side saying "Slow etcd operation(WalkWithPagination)" always with the same prefix "milvus-etcd/meta/datacoord-meta/s". Do you know what it is ?

```
[2023/10/24 14:40:38.080 +00:00] [WARN] [etcd/etcd_kv.go:648] ["Slow etcd operation(WalkWithPagination)"] ["time spent"=3.709275855s] [prefix=milvus-etcd/meta/datacoord-meta/s]
[2023/10/24 14:40:38.112 +00:00] [DEBUG] [config/etcd_source.go:141] ["etcd refreshConfigurations"] [prefix=milvus-etcd/config] [endpoints="[milvus-etcd:2379]"]
[2023/10/24 14:40:43.111 +00:00] [DEBUG] [config/etcd_source.go:141] ["etcd refreshConfigurations"] [prefix=milvus-etcd/config] [endpoints="[milvus-etcd:2379]"]
{"level":"warn","ts":"2023-10-24T14:40:44.372Z","logger":"etcd-client","caller":"v3@v3.5.5/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc00112e8c0/milvus-etcd:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
[2023/10/24 14:40:44.372 +00:00] [ERROR] [datacoord/server.go:571] ["retry func failed"] ["retry time"=20] [error="context deadline exceeded"] [stack="github.com/milvus-io/milvus/internal/datacoord.(*Server).initMeta
	/go/src/github.com/milvus-io/milvus/internal/datacoord/server.go:571
github.com/milvus-io/milvus/internal/datacoord.(*Server).initDataCoord
	/go/src/github.com/milvus-io/milvus/internal/datacoord/server.go:340
github.com/milvus-io/milvus/internal/datacoord.(*Server).Init
	/go/src/github.com/milvus-io/milvus/internal/datacoord/server.go:323
github.com/milvus-io/milvus/internal/distributed/datacoord.(*Server).init
	/go/src/github.com/milvus-io/milvus/internal/distributed/datacoord/service.go:125
github.com/milvus-io/milvus/internal/distributed/datacoord.(*Server).Run
	/go/src/github.com/milvus-io/milvus/internal/distributed/datacoord/service.go:248
github.com/milvus-io/milvus/cmd/components.(*DataCoord).Run
	/go/src/github.com/milvus-io/milvus/cmd/components/data_coord.go:50
github.com/milvus-io/milvus/cmd/roles.runComponent[...].func1
	/go/src/github.com/milvus-io/milvus/cmd/roles/roles.go:113"]
```

So I tried something : disable readiness/liveness probe to let the datacoord doing its stuff for a longer period of time (instead of crashing after 4min as asker by the liveness probe).
This allowed my datacoord to enter in a "Running" state  on the K8s side but after a moment it finally crash because I let hime retry its query until reaching 100 tries. I now have a new stacktrace maybe it could help.

[datacoord-logs.txt](https://github.com/milvus-io/milvus/files/13120365/datacoord-logs.txt)


Regards,

**Reply by yhmo** [reply]
Configuration doesn't help to reduce etcd latency.
The warning "Slow etcd operation(WalkWithPagination)" is thrown from here:
https://github.com/milvus-io/milvus/blob/53246b1b38f8fe25d5404b133ede62409237b400/internal/kv/etcd/etcd_kv.go#L99

In this function, it gets segments list information batch by batch by calling the etcd interface go.etcd.io/etcd/client/v3.Client.Get(). 3 seconds latency is far beyond expected. But I have no idea why this interface is so slow.

**Reply by Archalbc** [reply]
Hello,

I did network test between pods on same node and different nodes and was able to go far beyond the 10MB/s showed in my graph : 
* up to 5 Gbps node-to-node
* up to 50Gbps pod-to-pod  on the same node

Do you think of any software limitation on etcd or datacoord ? I can't find any on the internet...

Again I tested the disk and nothing wrong apparently.

While the datacoord is starting and loading stuff as you explained, ETCD CPU and Memory are quite low and without I/O wait.
I clearly don't understand why I get time_spent>3seconds and the "context deadline exceeded" error.

**Reply by yhmo** [reply]
No idea. You can try a local etcd to place the etcd on EBS. If the local etcd works well, then the problem is in the etcd on EBS.

**Reply by Archalbc** [reply]
Hi,

Today we were able to test using ebs-io1 with 15000 IOPS and the issue is still happening, that sound like a relief that we should not have to use such an expensive drive.
As we also exclude any network issue, we know think there is definitivly something wrong with the application (etcd or datacoord).

We need to understand what the datacoord try to do here : 
```
{"level":"warn","ts":"2023-10-24T14:40:44.372Z","logger":"etcd-client","caller":"v3@v3.5.5/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc00112e8c0/milvus-etcd:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
[2023/10/24 14:40:44.372 +00:00] [ERROR] [datacoord/server.go:571] ["retry func failed"] ["retry time"=20] [error="context deadline exceeded"] [stack="github.com/milvus-io/milvus/internal/datacoord.(*Server).initMeta
	/go/src/github.com/milvus-io/milvus/internal/datacoord/server.go:571
github.com/milvus-io/milvus/internal/datacoord.(*Server).initDataCoord
	/go/src/github.com/milvus-io/milvus/internal/datacoord/server.go:340
github.com/milvus-io/milvus/internal/datacoord.(*Server).Init
	/go/src/github.com/milvus-io/milvus/internal/datacoord/server.go:323
github.com/milvus-io/milvus/internal/distributed/datacoord.(*Server).init
	/go/src/github.com/milvus-io/milvus/internal/distributed/datacoord/service.go:125
github.com/milvus-io/milvus/internal/distributed/datacoord.(*Server).Run
	/go/src/github.com/milvus-io/milvus/internal/distributed/datacoord/service.go:248
github.com/milvus-io/milvus/cmd/components.(*DataCoord).Run
	/go/src/github.com/milvus-io/milvus/cmd/components/data_coord.go:50
github.com/milvus-io/milvus/cmd/roles.runComponent[...].func1
	/go/src/github.com/milvus-io/milvus/cmd/roles/roles.go:113"]
```

**Reply by xiaofan-luan** [reply]
could you try etcd cli to access your etcd? seems that the datacoord can not access etcd for some reason.
Did you make sure your etcd is still alive?

**Reply by Archalbc** [reply]
I can access etcd from datacoord pod with a "curl milvus-etcd:2379/health"
moreover, I can see this log at the datacoord boot :
 `[2023/10/30 09:20:41.248 +00:00] [INFO] [datacoord/server.go:560] ["data coordinator successfully connected to metadata store"] [metaType=etcd]
`

**Reply by xiaofan-luan** [reply]
so you can use etcdcli to access etcd, but the reboot failed?

There might be too much metas in etcd? how large is the cluster? @Archalbc 

@LoveEachDay is there way we can wait longer when start? what parameter did we need to change?

**Reply by Archalbc** [reply]
Yes I can even put/del key from another pod using `etcdctl --endpoints=milvus-etcd:2379 put foo bar`.
Here is my etcd node info
![image](https://github.com/milvus-io/milvus/assets/144429952/f757f349-3a93-447d-af26-12f1dd11519f)

I tried to read the code and it seems I'm in this loop : 
https://github.com/milvus-io/milvus/blob/master/internal/datacoord/server.go#L562

A little deeper, I guess something block me here  ? 
https://github.com/milvus-io/milvus/blob/master/internal/datacoord/meta.go#L103C24-L103C24
next : https://github.com/milvus-io/milvus/blob/master/internal/metastore/kv/datacoord/kv_catalog.go#L62
next : https://github.com/milvus-io/milvus/blob/master/internal/metastore/kv/datacoord/kv_catalog.go#L155
next: https://github.com/milvus-io/milvus/blob/master/internal/kv/etcd/etcd_kv.go#L66
finally my error logs : https://github.com/milvus-io/milvus/blob/master/internal/kv/etcd/etcd_kv.go#L99

Is this the timeout parameter ?
https://github.com/milvus-io/milvus/blob/master/internal/kv/etcd/etcd_kv.go#L69
https://github.com/milvus-io/milvus/blob/master/internal/kv/etcd/etcd_kv.go#L38
10 seconds and it is hard coded.

And the pod will finally crash once it reach connMetaMaxRetryTime here : 
https://github.com/milvus-io/milvus/blob/master/internal/datacoord/server.go#L571C61-L571C61

But even if we increase this timeout, I guess this will not explain why I'm reaching this issue while the cluster is not that heavy loaded


I read somewhere that indexcoord is merged with datacoord. Does it mean I need to remove the indexcoord right now ? Nothing has been done automatically if this is the case.

**Reply by LoveEachDay** [reply]
@Archalbc Which milvus version do you use in this case?

**Reply by Archalbc** [reply]
I'm now using 2.3.2 on EKS deployed with Helm.

**Reply by xiaofan-luan** [reply]
I think this this is related
https://github.com/milvus-io/milvus/pull/28026

But could you confirm how many keys are there in your etcd?
Seems that your binlog path numbers are incredibly higher than expected.
1. how many entities do you have?
2. Did you call frequent flush?

**Reply by xiaofan-luan** [reply]
Anyway good catch and really appreciate for your info.
We will release 2.3.3 to fix or you can use day build to fix

**Reply by Archalbc** [reply]
ho ho that's sound interesting !

Following my monitoring, we have 2M keys in the etcd clusters (I don't know if it's big).
1. how many entities do you have? => What do you mean by entities ?
2. Did you call frequent flush? => Let me ask to my backend developpers :D I'll answer to you asap.

I have just reseted the entire cluster (since I did a lot of changes/tests to fix this issue).
I'll re-activate my vectors injectors to reproduce the issue and try your fix for sure.

**Reply by xiaofan-luan** [reply]
2M seems to be a reasonable number.
I think the quick fix should work under your case.
We will release 2.3.3 soon

**Comment by xiaofan-luan** [comment]
> ho ho that's sound interesting !
> 
> Following my monitoring, we have 2M keys in the etcd clusters (I don't know if it's big).
> 
> 1. how many entities do you have? => What do you mean by entities ?
> 2. Did you call frequent flush? => Let me ask to my backend developpers :D I'll answer to you asap.
> 
> I have just reseted the entire cluster (since I did a lot of changes/tests to fix this issue). I'll re-activate my vectors injectors to reproduce the issue and try your fix for sure.

entitty is a milvus concept, on another word, how many vectors you stored ?

**Reply by Archalbc** [reply]
Just checked this morning after reset the DB yesterday.
We have reaching about 900K vectors  (my problem occur around this value already), and we want currently 1.7M key in etcd.
You will tell me, but I think it's pretty strange isn't it  ?
We plan to have around 120M vectors.

Backend developpers told me they call flush() after every insert() of maximum batch of 128 vectors.

**Reply by xiaofan-luan** [reply]
It doesn't make sense to have only 900k vectors but with 1.7M key in etcd.

could you share your code?

**Reply by xiaofan-luan** [reply]
Even with 120M vectors you shouldn't have 1.7M keys

**Reply by Archalbc** [reply]
Hello,
That's very interesting to know an idea of the number of keys in ETCD regarding the number of entities.
Thank to that, we pointed out with my developers some improvments that should be tested asap.
In case the situation doesn't improve, we will find a way to share with you the code.

Maybe it's a good idea to add this kind of information in the Milvus documentation ? Some tips about how to check if the developed workflow is correct. In my case, if I knew the proportion of keys in ETCD is abnormal regarding the number of vectors I would probably check the code instead of investigating the infrastructure.

Regards,

**Reply by Archalbc** [reply]
I confirm you fix improve the situation : the datacoord can now start and be in a Running state allowing vectors to be ingested again. I'm not sure though that datacoord is able to recover existing metadatas as it can't do its WalkWithPagination ;)
Errors in log still remains as the main issue is not resolved (slow operations...), I hope this will be fixed once we improve the code.

**Comment by Archalbc** [comment]
Hello,

I'm glad to announce that our issue is fixed !
We are now at 8,5M vectors for only around 700 keys on the ETCD clusters, the datacoord can now restart peacefully.

The fix: don't call flush() in your code, let Milvus do it by itself (every 1 seconds by default), unless it's very necessary (like a lot of UPSERT or unit testing).

I think documentation should be improved to add some stats between the number of key in ETCD regarding the number of vectors.
ETCD logs can be misleading as it will say stuff like "apply took too long" or "slow operation", you may think there is a disk performance issue, but that's was not the issue, the real one was the "way" we used Milvus.

Thank you again all for your help.

**Comment by xiaofan-luan** [comment]
> Hello,
> 
> I'm glad to announce that our issue is fixed ! We are now at 8,5M vectors for only around 700 keys on the ETCD clusters, the datacoord can now restart peacefully.
> 
> The fix: don't call flush() in your code, let Milvus do it by itself (every 1 seconds by default), unless it's very necessary (like a lot of UPSERT or unit testing).
> 
> I think documentation should be improved to add some stats between the number of key in ETCD regarding the number of vectors. ETCD logs can be misleading as it will say stuff like "apply took too long" or "slow operation", you may think there is a disk performance issue, but that's was not the issue, the real one was the "way" we used Milvus.
> 
> Thank you again all for your help.

Great! thanks for the feedback.

Yes, frequent flush will cause many keys in etcd and thus bring trouble in meta.

Thanks for the feedback, I will add more verification on the docuemnt. Enjoy!
